core/server/proxy/src/main/java/alluxio/proxy/s3/S3Handler.java

```diff
@@ -0,0 +1,518 @@
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    Matcher bucketMatcher = BUCKET_PATH_PATTERN.matcher(path);
+    Matcher objectMatcher = OBJECT_PATH_PATTERN.matcher(path);
+
+    String pathStr = path;
+    String bucket = null;
+    String object = null;
+    S3Handler handler = null;
+    try {
+      if (bucketMatcher.matches()) {
+        pathStr = path.substring(S3RequestServlet.S3_V2_SERVICE_PATH_PREFIX.length() + 1);
+        bucket = URLDecoder.decode(pathStr, "UTF-8");
+      } else if (objectMatcher.matches()) {
+        pathStr = path.substring(S3RequestServlet.S3_V2_SERVICE_PATH_PREFIX.length() + 1);
+        bucket = URLDecoder.decode(
+          pathStr.substring(0, pathStr.indexOf(AlluxioURI.SEPARATOR)), "UTF-8");
+        object = URLDecoder.decode(
+          pathStr.substring(pathStr.indexOf(AlluxioURI.SEPARATOR) + 1), "UTF-8");
+      }
+      handler = new S3Handler(bucket, object, request, response);
+      handler.setStopwatch(stopwatch);
+      handler.init();
+      S3BaseTask task = null;
+      if (object != null && !object.isEmpty()) {
+        task = S3ObjectTask.Factory.create(handler);
+      } else {
+        task = S3BucketTask.Factory.create(handler);
+      }
+      handler.setS3Task(task);
+      return handler;
+    } catch (Exception ex) {
+      LOG.error("Exception during create s3handler:{}", ThreadUtils.formatStackTrace(ex));
+      throw ex;
+    }
+  }
+
+  /**
+   * Process the response returned from S3Task core logic to write to downstream.
+   * @param servletResponse
+   * @param response
+   * @throws IOException
+   */
+  public static void processResponse(HttpServletResponse servletResponse,
+                                     Response response) throws IOException {
+    try {
+      // Status
+      servletResponse.setStatus(response.getStatus());
+      // Headers
+      final MultivaluedMap<String, String> headers = response.getStringHeaders();
+      for (final Map.Entry<String, List<String>> e : headers.entrySet()) {
+        final Iterator<String> it = e.getValue().iterator();
+        if (!it.hasNext()) {
+          continue;
+        }
+        final String header = e.getKey();
+        if (servletResponse.containsHeader(header)) {
+          // replace any headers previously set with values from Jersey container response.
+          servletResponse.setHeader(header, it.next());
+        }
+        while (it.hasNext()) {
+          servletResponse.addHeader(header, it.next());
+        }
+      }
+      // Entity
+      if (response.hasEntity()) {
+        ServletOutputStream servletOut = servletResponse.getOutputStream();
+        Object entity = response.getEntity();
+        if (entity instanceof InputStream) {
+          InputStream is = (InputStream) entity;
+          byte[] bytesArray = TLS_BYTES.get();
+          int read;
+          do {
+            try {
+              read = is.read(bytesArray);
```

Comment:
Does it affect the performance of the proxy?

Reply:
as discussed offline, this is doing the exact thing jetty does of handling read from the Response.entity(a inputstream in reading) and write to downstream httpchannel. Jetty's processing of response sits in org.glassfish.jersey.message.internal.writeTo(InputStream in, OutputStream out) which is also just a simple reading from inputstream and outstream. This is simply doing the samething with a tls bytes as transfer medium. We can do a lot more here in the future. There should not be any degradation of performance for that matter, but this gives us complete control in io manipulation and there's a lot of room for improvement starting from here.

Reply:
single linux machine with all components deployment, warp test result comparison:
warp get --host=localhost:16000 --access-key=alluxio -
-secret-key=alluxio --autoterm --bucket=test1 --duration=10m --concurrent 20      
[original route]:          
Operation: PUT
* Average: 73.98 MiB/s, 7.40 obj/s
Operation: GET
* Average: 214.00 MiB/s, 21.40 obj/s
[rearchitecture route]:
Operation: PUT
* Average: 73.28 MiB/s, 7.33 obj/s
Operation: GET
* Average: 218.19 MiB/s, 21.82 obj/s
[rearchitecture route with async processing turned on]:
Operation: PUT
* Average: 76.00 MiB/s, 7.60 obj/s
Operation: GET
* Average: 215.20 MiB/s, 21.52 obj/s

both sync / async processing with new architectures shows no degradation compared to original route.